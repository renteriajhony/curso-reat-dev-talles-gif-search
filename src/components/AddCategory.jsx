import { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ onAddCategory }) => {
  const [inputValue, setInputValue] = useState('');

  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const onSubmit = (event) => {
    event.preventDefault();
    const newValue = inputValue.trim();
    if (newValue.length <= 1) return;
    onAddCategory(newValue);
    setInputValue('');
  };
  return (
    <form onSubmit={onSubmit} aria-label='form'>
      <input
        type="text"
        placeholder="Search gifs..."
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onAddCategory: PropTypes.func.isRequired,
};
