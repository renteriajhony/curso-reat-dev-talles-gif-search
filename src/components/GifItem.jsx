import PropTypes from 'prop-types';

export const GifItem = ({ title, url }) => {
  return (
    <div className="card">
      <img key={url} src={url} alt={title} width={150} height={300} />
      <p key={title}>{title}</p>
    </div>
  );
};

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
