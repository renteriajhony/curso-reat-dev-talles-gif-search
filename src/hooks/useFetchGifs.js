import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (category) => {
  const [gifs, setGifs] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  const getImages = async () => {
    const images = await getGifs(category);
    setGifs(images);
    setIsLoading(false);
  };

  useEffect(() => {
    getImages();
  }, []);

  return {
    gifs,
    isLoading,
  };
};
