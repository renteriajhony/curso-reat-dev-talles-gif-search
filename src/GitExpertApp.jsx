import { useState } from 'react';
import { AddCategory, GifGrid } from './components';

export const GitExpertApp = () => {
  const [categories, setCategorie] = useState(['One Punch Man']);

  const onAddCategory = (category) => {
    categories.find(
      (item) => item.toLowerCase().trim() === category.toLowerCase().trim(),
    )
      ? alert('Category already exists!')
      : setCategorie([category, ...categories]);
  };

  return (
    <>
      <h1>GitExpertApp</h1>
      <AddCategory onAddCategory={onAddCategory} />

      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  );
};
