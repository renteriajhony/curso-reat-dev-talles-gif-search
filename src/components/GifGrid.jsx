import PropTypes from 'prop-types';
import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({ category }) => {
  const { gifs, isLoading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {isLoading && <p>Loading...</p>}
      <div className="card-grid" key={category}>
        {gifs.map((gif) => (
          <GifItem key={gif.id} {...gif} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
