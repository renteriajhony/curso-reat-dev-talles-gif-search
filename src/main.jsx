import React from 'react';
import ReactDOM from 'react-dom/client';
import { GitExpertApp } from './GitExpertApp.jsx';
import './styles.css';

ReactDOM.createRoot(document.getElementById('root')).render(
  //TODO: para producción, usar React.StrictMode, que se encarga de detectar errores de diseño, y manejarlos de manera segura
  // <React.StrictMode>
  //   <GitExpertApp />
  // </React.StrictMode>,

  <GitExpertApp />,
);
