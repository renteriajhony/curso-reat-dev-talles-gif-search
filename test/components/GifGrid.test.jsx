import { render, screen } from '@testing-library/react';
import { GifGrid } from '../../src/components';
import { useFetchGifs } from '../../src/hooks/useFetchGifs';

jest.mock('../../src/hooks/useFetchGifs');
describe('Pruebas en <GifGrid/>', () => {
  const category = 'Goku';
  const isLoading = true;

  test('debe hacer match con snapshop ', () => {
    useFetchGifs.mockImplementation(() => ({
      gifs: [],
      isLoading: true,
    }));
    const { container } = render(<GifGrid category={category} />);
    expect(container).toMatchSnapshot();
  });

  test('debe mostrar le loading inicialmente', () => {
    useFetchGifs.mockImplementation(() => ({
      gifs: [],
      isLoading: true,
    }));

    render(<GifGrid category={category} />);
    expect(screen.getByText('Loading...')).toBeTruthy();
    expect(screen.getByText(category)).toBeTruthy();
  });
  test('debe mostrar items cuando se cargan las imagenes useFetchGifs', () => {
    useFetchGifs.mockImplementation(() => ({
      gifs: [
        { id: 'ABC', title: 'Goku', url: 'https://localhost/goku' },
        { id: 'DEF', title: 'Vegeta', url: 'https://localhost/vegeta' },
      ],
      isLoading: false,
    }));

    render(<GifGrid category={category} />);
    expect(screen.getAllByRole('img').length).toBe(2);
  });
});
