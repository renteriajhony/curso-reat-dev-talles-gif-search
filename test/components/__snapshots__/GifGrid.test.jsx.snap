// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid/> debe hacer match con snapshop  1`] = `
<div>
  <h3>
    Goku
  </h3>
  <p>
    Loading...
  </p>
  <div
    class="card-grid"
  />
</div>
`;
